man/deva-compare-boundaries.1
.TH DEVA-VISIBILITY 1 "18 May 2018" "DEVA Project"
.SH NAME
deva-compare-boundaries \- compare one set of boundaries to another
.SH SYNOPSIS
\fBdeva-compare-boundaries\fR [\fIoptions\fR] \fIstandard.png\fR
\fIcomparison.png\fR \fIcoordinates\fR \fIvisualization.png\fR
.SH DESCRIPTION
Compares a binary image presumed to represent the location of real
boundary elements (e.g., from geometry) to a comparison binary image
presumed to represent the location of boundary elements to be compared
to the standard (e.g., luminance values).  The output is a visualization
of the distance, in visual angle, from each real boundary element to the
nearest comparison boundary element.
.PP
.SH OPTIONS
.TP
\fB\-\-red\-green\fR
Larger distance (usually associated with estimated visibility hazards)
shown in red and other geometric standard boundary elements that are
closer (usually associated with elements estimated to be less of a
potential visibility hazard) shown in green. Default.
.TP
\fB\-\-red\-gray\fR
Larger distance (usually associated with estimated visibility hazards)
shown in red and other geometric standard boundary elements that are
closer (usually associated with elements estimated to be less of a
potential visibility hazard) shown in dark gray.
.TP
\fB\-\-Gaussian=\fI<sigma>\fR
Visualization based on distance weighted by an unnormalized Gaussian
function with standard deviation \fI<sigma>\fR. Default visualization
weighting is \fB\-\-Gaussian=0.75\fR.
.TP
\fB\-\-reciprocal=\fI<scale>\fR
Visualization based on reciprocal of distance, with distance scaled by
\fI<scale>\fR.
.TP
\fB\-\-linear=\fI<max>\fR
Visualization linearly scaled to a maximum distance of \fI<max>\fR.
.TP
\fB\-\-quantscore\fR
Annotate the output image with the average hazard value over all of the
boundary elements in the standard.  Only available if compiled with the
Cairo library option.
.TP
\fB\-\-mask=\fI<filename.png>\fR
Name of existing PNG file where pixels that are TRUE should be marked
specially in the output visualization file.  This is usually the output
file from deva-visualize-geometry that is generated when the
\fB\-\-mask\fR flag is specified for that program.
.SH ARGUMENTS
.TP
\fIstandard.png\fR
Binary image file, with \fITRUE\fR indicated by non-zero value.
Represents the standard to compare against (e.g., geometry).
.TP
\fIcomparison.png\fR
Binary image file, with \fITRUE\fR indicated by non-zero value.
Represents the boundary elements to be compare to the standard (e.g.,
luminance boundaries).
.TP
\fIcoordinates\fR
A two line text file.  The first line specifies the units for the
\fIxyz.txt\fR and \fIdist.txt\fR files. The second line is the same as
the VIEW record in \fIinput.hdr\fR.  See \fBmake-coordinates-file\fR for
information on how to create this file.
.TP
\fIvisualization.png\fR
Output visualization file.
.\" SH EXAMPLES
\." SH LIMITATIONS
\." PP
.SH AUTHOR
William B. Thompson


CHANGES
version 3.3.00

Most fatal errors now print out the line number and file where the error
occurred.

Added option in deva-visibility to check for false positive luminance
boundary detections.

Added feature in deva-visibility (deva-commandline) and
deva-compare-boundaries to check for very dark regions in original
image.  This is motivated by a lack of knowledge about the effects of
illumination levels below about 1 cd/m^2 on low vision acuity and
contrast sensitivity.

FFT-based Gaussian blur routine added (deva-gblur-fft). This is used by
the code that checks for very dark regions.

Added routines in deva-margin for DEVA_float image object, providing
support needed by code that finds very dark regions in original image.

Added code in visualize-hazards for displaying very dark regions and
false positive errors in hazard visualization.

Added code supporting region-of-interest (ROI) processing in
determination of Visibility Hazard Scores in deva-visibility.

Changed contrast sensitivity of presets based on personal communication
(G.E. Legge & Y-Z Xiong, Feb. 19, 2019).

Added a "legalblind" preset to the other low vision categories in
deva-filter and deva-visibility.  This specifies the U.S.
"legal blindness" acuity level along with appropriate contrast
sensitivity and color sensitivity values.  Note that this specifies a
level of visual impairment right at the legal blindness cutoff point.
It is not some sort of test of legal blindness.

Added a "normal" preset to the other low vision categories in
deva-filter and deva-visibility.  This is treated as a silent option,
not included in the prompt or man page, because the filter has not been
tested for this condition.

Cleaned up where new image objects were allocated in deva-visibility.c
and related routines.

Minor comment cleanup.

Cleanups and additions to man pages.

Various changes to CMakeLists.txt

Ran this file through a spell checker.

version 3.2.04

luminance-boundaries now issues a warning to avoid use with input images
generated by deva-filter.  The problem may be due to quantization noise
in the low precision Radiance format input image resulting in a large
number of micro edges that in turn confuse the auto thresholding
routine.

version 3.2.03

Really fix deva-visibility man page.

version 3.2.02

Minor fix to deva-visibility man page.

version 3.2.01

Added build process for MacOS and standardized directory layout for
Windows and MacOS builds.

Removed extraneous debugging command from deva-margin.c

version 3.2.00

Added two new programs: luminance-boundaries, geometry-boundaries.

In deva-filter.c (deva-filter, deva-visibility): Changed smoothing
algorithm to better deal with artifacts at the boundaries of the
smoothing windows.  Added optional debugging code to output individual
contrast bands.

In deva-margin.c:  Added compile-time option to base feathering on
average luminance over full image rather than just over the boundary
pixels.  Added compile-time option to base margin pixels on reflection
of image around the boundary, not just a fill of boundary pixels.

In dilate.c, added dt_euclid_sq_2 which used a pre-allocated output
image.

In deva-commandline.c (deva-filter, deva-visibility): Cleaned up command
summaries.  Allowed specifying smoothing_type with presets.

In ChungLeggeCSF.h: Cleanup of unused ChungLeggeCSF_PEAK_FREQUENCY
defines.

Man pages for new programs, minor updates to existing man pages.

version 3.1.02

In deva-commandline.c (deva-visibility): Added measurement type options
to usage message.

version 3.1.01

deva-compare-boundaries.c: New program to create a visualization image
comparing a standard set of boundary elements (e.g., geometry) to a
different set of boundary elements (e.g., luminance edges).  Annotation
of output image, which requires Cairo library, is includable with a
compile-time switch.

deva-add-text.c: New routine to add annotation text to an image.
Requires a switch in CMakeLists.txt to link against Cairo library.

In deva-commandline.c (deva-filter, deva-visibility): Switch to basing
auto-clipping on multiple of median luminance, which is more robust than
the previous multiple of average luminance.  Added a compile-time option
for basing margin on only smaller dimension of input image.

In deva-commandline.c (deva-visibility): Added interface for more
measurement types.

In deva-canny.c:  Switch to basing luminance boundaries on log
luminance, not luminance itself.  This better approximates human
perception of contrast boundaries.  Switched to basing auto-thresholding
on local maxima of directional gradient, not all directional gradients.

In visualize-hazards.c/visualize-hazards.h: Cleaned up calling sequence.
Added in more measurement types.
 
In read-geometry.c, fixed possibly uninitialized DEVA_null_view.

Additional commenting.

version 3.0.01

Modified radiance-header.c to deal with indented VIEW records produced
by pcomb.  (pcomb has been modified in development version of Radiance
to output one non-indented VIEW record, alleviating the need for this
change if a recent enough version of Radiance is used to generate the
file.)

Cleaned up deva-gblur.c.  There should be no change in functionality.

Better error reporting in read-geometry routines.

Removed one extraneous routine from deva-margin.c

More extensive comments throughout.

Minor updates to man pages.

version 3.0.00

Added deva-visibility.  This extends the functionality of deva-filter to
provide estimates of likely low-vision hazards, defined as geometric
structures that may be a mobility hazard but are not co-located with
detectable visual contrast.  deva-visibility shares much of the
deva-filter code base.

Added deva-visualize-geometry, a utility program to display geometry
files used by deva-visibility.

Added make-coordinates-file, a support routine used to create one of the
input files needed by deva-visibility.

Added a build for Windows (see INSTALLATION file).  This requires a

deva-compare-boundaries.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <libgen.h>
#include <strings.h>
#include <string.h>
#include "dilate.h"
#include "deva-visibility.h"
#include "visualize-hazards.h"
#include "deva-image.h"
#ifdef DEVA_USE_CAIRO
#include "deva-add-text.h"
#endif	/* DEVA_USE_CAIRO */
#include "DEVA-png.h"
#include "read-geometry.h"

#define	HAZARD_NO_EDGE	(-1.0)
#define	TEXT_DEFAULT_FONT_SIZE	24.0

#define	SQ(x)	((x) * (x))

char	*Usage =
"deva-compare-boundaries [--red-green|--red-gray]"
	"\n\t[--Gaussian=<sigma>|--reciprocal=<scale>|--linear=<max>]"
#ifdef DEVA_USE_CAIRO
	"\n\t[--quantscore] [--fontsize=<n>]"
#endif	/* DEVA_USE_CAIRO */
	"\n\t[--mask=<mask-filename>]"
	"\n\tstandard.png comparison.png coord visualization.png";
int	args_needed = 4;

int		    imax ( int a, int b );
DEVA_float_image    *compute_hazards ( DEVA_gray_image *standard,
			DEVA_float_image *comparison_edge_distance,
			double degrees_per_pixel );
void		    make_visible ( DEVA_gray_image *boundaries );

#ifdef DEVA_USE_CAIRO
void		    add_quantscore ( char *standard_name,
			char *comparison_name, DEVA_float_image *hazards,
       			DEVA_RGB_image *hazards_visualization,
			double text_font_size,
			Measurement_type measurement_type,
			double scale_parameter,
			Visualization_type visualization_type,
			int log_output_flag, char *log_output_filename );
#endif	/* DEVA_USE_CAIRO */

int
main ( int argc, char *argv[] )
{
#ifdef DEVA_USE_CAIRO
    int			log_output_flag = FALSE;
    char		*log_output_filename = NULL;
#endif	/* DEVA_USE_CAIRO */
    Visualization_type	visualization_type = red_gray_type;	/* default */
    Measurement_type	measurement_type = Gaussian_measure;	/* default */
    double		max_hazard = 2.0;			/* default */
    double		reciprocal_scale = 1.0;			/* default */
    double		Gaussian_sigma = 0.75;			/* default */
    double		scale_parameter;

#ifdef DEVA_USE_CAIRO
    int			quantscore = FALSE;
    double		text_font_size = TEXT_DEFAULT_FONT_SIZE;
#endif	/* DEVA_USE_CAIRO */

    DEVA_gray_image	*standard;
    char		*standard_name;
    DEVA_gray_image	*comparison;
    char		*comparison_name;
    DEVA_coordinates	*coordinates;
    double		degrees_per_pixel;
    DEVA_float_image	*comparison_edge_distance;
    DEVA_float_image	*hazards;
    DEVA_RGB_image	*hazards_visualization;
    char		*mask_file_name = NULL;
    DEVA_gray_image	*mask = NULL;
    int			argpt = 1;

    while ( ( ( argc - argpt ) >= 1 ) && ( argv[argpt][0] == '-' ) ) {
	if ( ( strcasecmp ( argv[argpt], "--red-green" ) == 0 ) ||
		( strcasecmp ( argv[argpt], "-red-green" ) == 0 ) ) {
	    visualization_type = red_green_type;
	    argpt++;
	} else if ( ( strcasecmp ( argv[argpt], "--red-gray" ) == 0 ) ||
		( strcasecmp ( argv[argpt], "-red-gray" ) == 0 ) ) {
	    visualization_type = red_gray_type;
	    argpt++;
#ifdef DEVA_USE_CAIRO
	} else if ( ( strcasecmp ( argv[argpt], "--quantscore" ) == 0 ) ||
		( strcasecmp ( argv[argpt], "-quantscore" ) == 0 ) ) {
	    quantscore = TRUE;
	    argpt++;
	} else if ( strncasecmp ( argv[argpt], "--fontsize=",
		    strlen ( "--fontsize=" ) ) == 0 ) {
	    text_font_size = atof ( argv[argpt] +
		    strlen ( "--fontsize=" ) );
	    argpt++;
	} else if ( strncasecmp ( argv[argpt], "-fontsize=",
		    strlen ( "-fontsize=" ) ) == 0 ) {
	    text_font_size = atof ( argv[argpt] +
		    strlen ( "-fontsize=" ) );
	    argpt++;
#endif	/* DEVA_USE_CAIRO */
	} else if ( strncasecmp ( argv[argpt], "--reciprocal=",
		    strlen ( "--reciprocal=" ) ) == 0 ) {
	    measurement_type = reciprocal_measure;
	    reciprocal_scale = atof ( argv[argpt] +
		    strlen ( "--reciprocal=" ) );
	    argpt++;
	} else if ( strncasecmp ( argv[argpt], "-reciprocal=",
		    strlen ( "-reciprocal=" ) ) == 0 ) {
	    measurement_type = reciprocal_measure;
	    reciprocal_scale = atof ( argv[argpt] +
		    strlen ( "-reciprocal=" ) );
	    argpt++;
	} else if ( strncasecmp ( argv[argpt], "--linear=",
		    strlen ( "--linear=" ) ) == 0 ) {
	    measurement_type = linear_measure;
	    max_hazard = atof ( argv[argpt] + strlen ( "--linear=" ) );
	    argpt++;
	} else if ( strncasecmp ( argv[argpt], "-linear=",
		    strlen ( "-linear=" ) ) == 0 ) {
	    measurement_type = linear_measure;
	    max_hazard = atof ( argv[argpt] + strlen ( "-linear=" ) );
	    argpt++;
	} else if ( strncasecmp ( argv[argpt], "--Gaussian=",
		    strlen ( "--Gaussian=" ) ) == 0 ) {
	    measurement_type = Gaussian_measure;
	    Gaussian_sigma = atof ( argv[argpt] + strlen ( "--Gaussian=" ) );
	    argpt++;
	} else if ( strncasecmp ( argv[argpt], "-Gaussian=",
		    strlen ( "-Gaussian=" ) ) == 0 ) {
	    measurement_type = Gaussian_measure;
	    Gaussian_sigma = atof ( argv[argpt] + strlen ( "-Gaussian=" ) );
	    argpt++;

	} else if ( strncasecmp ( argv[argpt], "--mask=",
		    strlen ( "--mask=" ) ) == 0 ) {
	    mask_file_name = argv[argpt] + strlen ( "--mask=" );
	    argpt++;

	} else if ( strncasecmp ( argv[argpt], "-mask=",
		    strlen ( "-mask=" ) ) == 0 ) {
	    mask_file_name = argv[argpt] + strlen ( "-mask=" );
	    argpt++;

	/* hidden options */
#ifdef DEVA_USE_CAIRO
	} else if ( strncasecmp ( argv[argpt], "--log=",
		    strlen ( "--log=" ) ) == 0 ) {
	    log_output_flag = TRUE;
	    log_output_filename = argv[argpt] + strlen ( "--log=" );
	    argpt++;
	} else if ( strncasecmp ( argv[argpt], "-log=",
		    strlen ( "-log=" ) ) == 0 ) {
	    log_output_flag = TRUE;
	    log_output_filename = argv[argpt] + strlen ( "-log=" );
	    argpt++;
#endif	/* DEVA_USE_CAIRO */
	} else {
	    fprintf ( stderr, "deva-compare-boundaries: invalid flag (%s)!\n",
		    argv[argpt] );
	    return ( EXIT_FAILURE );	/* error return */
	}
    }

    if ( ( argc - argpt ) != args_needed ) {
	fprintf ( stderr, "%s\n", Usage );
	return ( EXIT_FAILURE );        /* error return */
    }

    standard_name = argv[argpt++];
    standard = DEVA_gray_image_from_filename_png ( standard_name );
    comparison_name = argv[argpt++];
    comparison = DEVA_gray_image_from_filename_png ( comparison_name );

    if ( !DEVA_image_samesize ( standard, comparison ) ) {
	fprintf ( stderr, "%s and %s not same size!\n", standard_name,
		comparison_name );
	DEVA_print_file_lineno ( __FILE__, __LINE__ );
	return ( EXIT_FAILURE );        /* error return */
    }

    coordinates = DEVA_coordinates_from_filename ( argv[argpt++] );

    degrees_per_pixel = fmax ( coordinates->view.vert,
	    coordinates->view.horiz ) /
	((double) imax ( DEVA_image_n_rows ( standard ),
	    DEVA_image_n_cols ( standard ) ) );

    /* squared-distance transform */
    comparison_edge_distance = dt_euclid_sq ( comparison );

    /*
     * Compute distance, represented as a visual angle, from each standard
     * boundary pixel to nearest comparison boundary pixel.
     */

    hazards = compute_hazards ( standard, comparison_edge_distance,
	    degrees_per_pixel );

    /* clean up */
    DEVA_gray_image_delete ( comparison );
    DEVA_float_image_delete ( comparison_edge_distance );

    switch ( measurement_type ) {
	case reciprocal_measure:
	    scale_parameter = reciprocal_scale;
	    break;

	case linear_measure:
	    scale_parameter = max_hazard;
	    break;

	case Gaussian_measure:
	    scale_parameter = Gaussian_sigma;
	    break;

	default:
	    fprintf ( stderr, "invalid measurement_type!\n" );
	    DEVA_print_file_lineno ( __FILE__, __LINE__ );
	    exit ( EXIT_FAILURE );
    }

    if ( mask_file_name != NULL ) {
	mask = DEVA_gray_image_from_filename_png ( mask_file_name );
    }

    hazards_visualization = visualize_hazards ( hazards, measurement_type,
	    scale_parameter, visualization_type, mask, NULL, &standard, NULL );

    /* clean up */
    DEVA_gray_image_delete ( standard );

#ifdef DEVA_USE_CAIRO
    if ( quantscore ) {
	add_quantscore ( standard_name, comparison_name, hazards,
		hazards_visualization, text_font_size,
		measurement_type, scale_parameter,
		visualization_type,
		log_output_flag, log_output_filename );
    }
#endif	/* DEVA_USE_CAIRO */

    DEVA_RGB_image_to_filename_png ( argv[argpt++], hazards_visualization );

    /* clean up */
    DEVA_float_image_delete ( hazards );
    DEVA_RGB_image_delete ( hazards_visualization );
    DEVA_coordinates_delete ( coordinates );
    if ( mask_file_name != NULL ) {
	DEVA_gray_image_delete ( mask );
    }

    return ( EXIT_SUCCESS );	/* normal exit */
}

int
imax ( int a, int b )
{
    if ( a > b ) {
	return ( a );
    } else {
	return ( b );
    }
}

DEVA_float_image *
compute_hazards ( DEVA_gray_image *standard,
	DEVA_float_image *comparison_edge_distance, double degrees_per_pixel )
/*
 * Compute distance, represented as a visual angle, from each geometric
 * boundary pixel to nearest luminance boundary pixel.
 *
 * standard:	TRUE is pixel is a geometric edge.
 *
 * comparison_edge_distance:	Squared distance to nearest luminance edge.
 *
 * degrees_per_pixel:	Angle in degrees between two horizontally or vertically
 * 			adjacent pixel locations.
 *
 * returned value:	DEVA_float_image image object. Non-negative pixel
 * 			values give distance from geometry boundary at
 * 			pixel location to nearest luminance boundary element.
 * 			Negative pixel values indicate that there is no
 * 			geometry boundary at that location.
 */
{
    int			row, col;
    DEVA_float_image	*hazards;

    if ( ! DEVA_image_samesize ( standard, comparison_edge_distance ) ) {
	fprintf ( stderr, "compute_hazards: argument size mismatch!\n" );
	DEVA_print_file_lineno ( __FILE__, __LINE__ );
	exit ( EXIT_FAILURE );
    }

    if ( degrees_per_pixel <= 0.0 ) {
	fprintf ( stderr, "compute_hazards: invalid degrees_per_pixel (%f)\n",
		degrees_per_pixel );
	DEVA_print_file_lineno ( __FILE__, __LINE__ );
	exit ( EXIT_FAILURE );
    }

    hazards = DEVA_float_image_new ( DEVA_image_n_rows ( standard ),
	    DEVA_image_n_cols ( standard ) );

    for ( row = 0; row < DEVA_image_n_rows ( standard ); row++ ) {
	for ( col = 0; col < DEVA_image_n_cols ( standard );
		col++ ) {
	    if ( DEVA_image_data ( standard, row, col ) ) {
		DEVA_image_data ( hazards, row, col ) = degrees_per_pixel *
		    sqrt ( DEVA_image_data ( comparison_edge_distance, row,
				col ) );
	    } else {
		DEVA_image_data ( hazards, row, col ) = HAZARD_NO_EDGE;
	    }
	}
    }

    return ( hazards );
}

void
make_visible ( DEVA_gray_image *boundaries )
/*
 * Change 0,1 values in a boolean file to 0,255.  This preserves TRUE/FALSE,
 * while making the file directly displayable.
 */
{
    int     row, col;

    for ( row = 0; row < DEVA_image_n_rows ( boundaries ); row ++ ) {
	for ( col = 0; col < DEVA_image_n_cols ( boundaries );
		col ++ ) {
	    if ( DEVA_image_data ( boundaries, row,
			col ) ) {
		DEVA_image_data ( boundaries, row, col ) = 255;
	    }
	}
    }
}

#define	BUFF_SIZE		200
#define	TEXT_ROW_1		30.0
#define	TEXT_ROW_2		60.0
#define	TEXT_ROW_3		90.0
#define	TEXT_COL		10.0
#define	TEXT_RED		1.0
#define	TEXT_GREEN		1.0
#define	TEXT_BLUE		1.0

#ifdef DEVA_USE_CAIRO

void
add_quantscore ( char *standard_name, char *comparison_name,
	DEVA_float_image *hazards, DEVA_RGB_image *hazards_visualization,
	double text_font_size,
	Measurement_type measurement_type, double scale_parameter,
	Visualization_type visualization_type,
	int log_output_flag, char *log_output_filename )
{
    int		n_rows, n_cols;
    int		row, col;
    double	sum;
    int		count;
    char	text[BUFF_SIZE];
    char	*visualization_code;

    DEVA_RGBf	text_color;
    cairo_surface_t *cairo_surface;
    FILE	*log_file_fd;

    n_rows = DEVA_image_n_rows ( hazards );
    n_cols = DEVA_image_n_cols ( hazards );

    sum = 0.0;
    count = 0;

    for ( row = 0; row < n_rows; row++ ) {
	for ( col = 0; col < n_cols; col++ ) {
	    if ( DEVA_image_data ( hazards, row, col ) >= 0 ) {

		/*
		 * For matching measures, high values are good, low values
		 * are bad.  (This is opposite the conventions for hazard
		 * visualization.)
		 */

		switch ( measurement_type ) {

		    case reciprocal_measure:
			sum += scale_parameter /
			    ( DEVA_image_data ( hazards, row, col ) +
			    				scale_parameter );
			break;

		    case linear_measure:
			sum += 1.0 - ( fmin ( DEVA_image_data ( hazards, row,
				col ), scale_parameter ) / scale_parameter );
			break;

		    case Gaussian_measure:
			sum += exp ( -0.5 * ( SQ ( DEVA_image_data ( hazards,
					    row, col ) / scale_parameter ) ) );
			break;

		    default:
			fprintf ( stderr,
				"deva-compare-boundaries: internal error!\n" );
			DEVA_print_file_lineno ( __FILE__, __LINE__ );
			exit ( EXIT_FAILURE );	/* error return */
			break;
		}

		count++;
	    }
	}
    }

    visualization_code = NULL;

    switch ( visualization_type ) {

	case red_gray_type:
	    visualization_code = "red-gray";
	    break;

	case red_green_type:
	    visualization_code = "red-green";
	    break;

	default:
	    fprintf ( stderr, "deva-compare-boundaries: internal error!\n" );
	    break;
    }

    cairo_surface = DEVA_RGB_cairo_open ( hazards_visualization );

    text_color.red = TEXT_RED;
    text_color.green = TEXT_GREEN;
    text_color.blue = TEXT_BLUE;

    if ( count == 0 ) {
	snprintf ( text, BUFF_SIZE,
		"%s / %s: no pixels in standard!",
		basename ( standard_name ), basename ( comparison_name ) );
	DEVA_cairo_add_text ( cairo_surface, TEXT_ROW_1, TEXT_COL,
		text_font_size, text_color, text );
    } else {

	snprintf ( text, BUFF_SIZE, "%s / %s:",
		basename ( standard_name ), basename ( comparison_name ) );
	DEVA_cairo_add_text ( cairo_surface, TEXT_ROW_1, TEXT_COL,
		text_font_size, text_color, text );

	switch ( measurement_type ) {
	    case reciprocal_measure:
		snprintf ( text, BUFF_SIZE,
			"reciprocal measure (scale = %.2f), %s",
			scale_parameter, visualization_code );
		break;

	    case linear_measure:
		snprintf ( text, BUFF_SIZE,
			"linear measure (max_hazard = %.2f), %s",
			scale_parameter, visualization_code );
		break;

	    case Gaussian_measure:
		snprintf ( text, BUFF_SIZE,
			"Gaussian measure (Gaussian_sigma = %.2f), %s",
			scale_parameter, visualization_code );
		break;

	    default:
		fprintf ( stderr,
			"deva-compare-boundaries: internal error!\n" );
		DEVA_print_file_lineno ( __FILE__, __LINE__ );
		exit ( EXIT_FAILURE );	/* error return */
		break;
	}

	DEVA_cairo_add_text ( cairo_surface, TEXT_ROW_2, TEXT_COL,
		text_font_size, text_color, text );

	snprintf ( text, BUFF_SIZE, "matching score = %.3f",
		sum / ((double) count ) );
	DEVA_cairo_add_text ( cairo_surface, TEXT_ROW_3, TEXT_COL,
		text_font_size, text_color, text );
    }

    DEVA_RGB_cairo_close_inplace ( cairo_surface, hazards_visualization );

    if ( log_output_flag ) {
	log_file_fd = fopen ( log_output_filename, "a" );
	if ( log_file_fd == NULL ) {
	    perror ( log_output_filename );
	    DEVA_print_file_lineno ( __FILE__, __LINE__ );
	    exit ( EXIT_FAILURE );
	}

	if ( count > 0 ) {
	    fprintf ( log_file_fd, ",%.3f", sum / ((double) count ) );
	} else {
	    fprintf ( log_file_fd, "," );
	}

	fclose ( log_file_fd );
    }
}

#endif	/* DEVA_USE_CAIRO */
